libopae/plugins/xfpga/metrics/vector.c

```diff
@@ -0,0 +1,182 @@
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+/**
+* \file vector.c
+* \brief fpga metrics vector
+*/
+
+#include "vector.h"
+#include "opae/access.h"
+#include "opae/utils.h"
+#include "opae/manage.h"
+#include "opae/enum.h"
+#include "opae/properties.h"
+#include "common_int.h"
+#include "safe_string/safe_string.h"
+
+#define FPGA_VECTOR_CAPACITY 20
+
+fpga_result fpga_vector_init(fpga_metric_vector *vector)
+{
+	fpga_result result = FPGA_OK;
+
+	if (vector == NULL)
+		return FPGA_INVALID_PARAM;
+
+	vector->capacity = FPGA_VECTOR_CAPACITY;
+	vector->total = 0;
+	vector->fpga_metric_item = malloc(sizeof(void *) * vector->capacity);
+
+	if (vector->fpga_metric_item == NULL)
+		return FPGA_NO_MEMORY;
+
+	return result;
+}
+
+fpga_result fpga_vector_free(fpga_metric_vector *vector)
+{
+	fpga_result result = FPGA_OK;
+	if (vector == NULL) {
+		FPGA_ERR("Invalid parm");
+		return FPGA_INVALID_PARAM;
+	}
+	free(vector->fpga_metric_item);
+
+	vector->fpga_metric_item = NULL;
+
+	return result;
+}
+
+uint64_t fpga_vector_total(fpga_metric_vector *vector)
+{
+	if (vector == NULL) {
+		FPGA_ERR("Invalid parm");
+		return 0;
+	}
+	return vector->total;
+}
+
+fpga_result fpga_vector_resize(fpga_metric_vector *vector, uint64_t capacity)
+{
+	fpga_result result = FPGA_OK;
+
+	if (vector == NULL) {
+		FPGA_ERR("Invalid parm");
+		return FPGA_INVALID_PARAM;
+	}
+
+	void **fpga_metric_item = realloc(vector->fpga_metric_item, sizeof(void *) * capacity);
+
+	if (fpga_metric_item == NULL) {
+		FPGA_ERR("Invalid parm");
+		return FPGA_NO_MEMORY;
+	}
+
+	if (fpga_metric_item) {
+		vector->fpga_metric_item = fpga_metric_item;
+		vector->capacity = capacity;
+	}
+	return result;
+
+}
+
+fpga_result fpga_vector_push(fpga_metric_vector *vector, void *fpga_metric_item)
+{
+	fpga_result result = FPGA_OK;
+
+	if ((vector == NULL) ||
+		(fpga_metric_item == NULL)) {
+		FPGA_ERR("Invalid parm");
+		return FPGA_INVALID_PARAM;
+	}
+
+	if (vector->capacity == vector->total) {
+
+		result = fpga_vector_resize(vector, vector->capacity * 2);
+		if (result != FPGA_OK)
+			return result;
+	}
+	vector->fpga_metric_item[vector->total++] = fpga_metric_item;
+
+
+	return result;
+
+}
+
+void *fpga_vector_pop(fpga_metric_vector *vector, uint64_t index)
```

Comment:
This routine isn't actually modifying anything within vector.  I question whether it should be called fpga_vector_pop.  How about fpga_vector_item?

Reply:
You have fpga_vector_get below.  Is this routine actually needed?